api/src/services/assessmentsDummyService.ts

```diff
@@ -0,0 +1,185 @@
+
+  let matchingAsssessmentSubmissionsRows: AssessmentSubmissionRow[] = await db(
+    'assessment_submissions'
+  )
+    .select('id', 'assessment_id')
+    .where({ assessment_id: assessmentId, principal_id: principalId });
+
+  if (matchingAsssessmentSubmissionsRows.length) {
+    // - if a row exists, update the status and submittedAt dates (if one is passed) and return the id of that row
+
+    await db(`assessment_submissions`)
+      .update({
+        assessment_submission_state_id: assessmentSubmissionStateId,
+        score: score,
+        opened_at: openedAt,
+        submitted_at: submittedAt,
+      })
+      .where({ assessment_id: assessmentId, principal_id: principalId });
+  } else {
+    // - if a row doesn't exist, create one and return the id of that row in the assessment_submissions table
+    await db(`assessment_submissions`).insert({
+      assessment_id: assessmentId,
+      principal_id: principalId,
+      assessment_submission_state_id: assessmentSubmissionStateId,
+      score: score,
+      opened_at: openedAt,
+      submitted_at: submittedAt,
+    });
+  }
+
+  matchingAsssessmentSubmissionsRows = await db<AssessmentSubmissionRow>(
+    'assessment_submissions'
+  )
+    .select('id')
+    .where({ assessment_id: assessmentId, principal_id: principalId });
+
+  for (const response of responses) {
+    await insertToAssessmentResponses(
+      assessmentId,
+      matchingAsssessmentSubmissionsRows[0].id,
+      response.question_id,
+      response.answer_id,
+      response.response,
+      response.score,
+      response.grader_response
+    );
+  }
+
+  const insertedRowParsed = {
+    id: matchingAsssessmentSubmissionsRows[0].id,
+    principal_id: principalId,
+    assessment_id: assessmentId,
+    score: score,
+    opened_at: openedAt,
+    submitted_at: submittedAt,
+    responses: responses,
+  };
+
+  return insertedRowParsed;
+};
+
+export const insertToAssessmentResponses = async (
+  assessmentId: number,
+  submissionId: number,
+  questionId: number,
+  answerId?: number,
+  response?: string,
+  score?: number,
+  graderResponse?: string
+) => {
+  const assessmentResponsesWithMatchedId = await db(`assessment_responses`)
+    .select('id', 'assessment_id')
+    .where({ assessment_id: assessmentId, submission_id: submissionId });
+  if (assessmentResponsesWithMatchedId.length) {
+    await db(`assessment_responses`)
+      .update({
+        answer_id: answerId,
+        response: response,
+        score: score,
+        grader_response: graderResponse,
+      })
+      .where({ assessment_id: assessmentId, submission_id: submissionId });
+  } else {
+    await db(`assessment_responses`).insert({
+      assessment_id: assessmentId,
+      submission_id: submissionId,
+      question_id: questionId,
+      answer_id: answerId,
+      response: response,
+      score: score,
+      grader_response: graderResponse,
+    });
+  }
+};
```

Comment:
You're missing the final `SELECT` here that will return the `id` of the row that was inserted.